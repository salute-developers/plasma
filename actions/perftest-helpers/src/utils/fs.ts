import * as fs from 'fs';

export async function readJson<T>(path: string): Promise<T> {
    return JSON.parse(await fs.promises.readFile(path, 'utf-8'));
}
