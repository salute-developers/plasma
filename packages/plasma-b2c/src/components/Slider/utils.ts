import { MutableRefObject } from 'react';

/**
 * Расчитать значение слайдера с учетом его координат и шага изменений.
 * @param {number} handleCenterXRelative
 * @param {number} stepSize
 * @param {number} min
 * @param {number} max
 * @return {number}
 */
export function getSliderThumbValue(handleCenterXRelative: number, stepSize: number, min: number, max: number) {
    const newValue = Math.round(handleCenterXRelative / stepSize) + min;

    return Math.min(Math.max(newValue, min), max);
}

/**
 * Расчитывает значение отступа слайдера на основе его положения (справа, слева) на отрезке слайдера.
 * Значение используется для правильного расчета ограничения движения слайдера.
 * @param ref
 * @param {'left' | 'right'} side
 * @return Array<number>
 */
export function getOffsets(
    ref: ((instance: HTMLDivElement | null) => void) | MutableRefObject<HTMLDivElement | null> | null,
    side?: 'left' | 'right',
): number[] {
    if (!ref || !('current' in ref) || !ref.current || !side) {
        return [0, 0];
    }

    const size = ref.current.clientWidth;

    if (side === 'left') {
        return [0, size];
    }

    if (side === 'right') {
        return [size, 0];
    }

    return [0, 0];
}
