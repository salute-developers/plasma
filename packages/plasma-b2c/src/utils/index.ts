export * from '@salutejs/plasma-web/utils';
