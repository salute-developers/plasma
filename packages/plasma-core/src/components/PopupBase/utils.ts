import { css } from "styled-components";
import { BasicPopupBasePlacement, PopupBasePlacement } from "./types";

export const handlePosition = (placement?: PopupBasePlacement, offset?: [number | string, number | string]) => {
    let x = '0rem';
    let y = '0rem';
    if (offset) {
        const [_x, _y] = offset;
        x = typeof _x === 'number' ? `${_x}rem` : _x;
        y = typeof _y === 'number' ? `${_y}rem` : _y;
    }

    if (!placement || placement === 'center') {
        return css`
            left: calc(50% + ${x});
            top: calc(50% - ${y});
            transform: translate(-50%, -50%);
        `;
    }

    let left;
    let right;
    let top;
    let bottom;
    const placements = placement.split('-') as BasicPopupBasePlacement[];

    placements.forEach((placement: BasicPopupBasePlacement) => {
        switch (placement) {
            case 'left':
                left = x;
                break;
            case 'right':
                right = x;
                break;
            case 'top':
                top = y;
                break;
            case 'bottom':
                bottom = y;
                break;
            default:
                break;
        }
    });

    const isCenteredX = left === undefined && right === undefined;
    const isCenteredY = top === undefined && bottom === undefined;

    return css`
        left: ${left};
        right: ${right};
        top: ${top};
        bottom: ${bottom};
        ${isCenteredX &&
        css`
            left: calc(50% + ${x});
            transform: translateX(-50%);
        `}
        ${isCenteredY &&
        css`
            top: calc(50% - ${y});
            transform: translateY(-50%);
        `}
    `;
};