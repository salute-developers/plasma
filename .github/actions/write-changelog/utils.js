import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';
import { promises as fs } from 'fs';

function getFilePath(pkg, fileName) {
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = dirname(__filename);

    return resolve(__dirname, '../../..', `packages/${pkg}/${fileName}`);
}

function getDate(date) {
    const formatter = new Intl.DateTimeFormat('ru', {
        day: 'numeric',
        month: 'long',
        year: 'numeric',
    });
    const data = formatter.format(date).replace(' г.', '');

    return `(${data})`;
}

async function getPackageVersion(pkg) {
    const filePath = getFilePath(pkg, 'package.json');

    try {
        const packageJson = await fs.readFile(filePath, 'utf8');
        const { version } = JSON.parse(packageJson);

        return `# ${version}`;
    } catch (error) {
        console.error('Error reading package.json:', error);
    }
}

export async function writeChangelog(markdown, pkg) {
    try {
        if (!pkg) {
            console.error('Не указан пакет для записи об изменениях');

            return;
        }

        const version = await getPackageVersion(pkg);
        const mdFilePAth = getFilePath(pkg, 'CHANGELOG.md');
        const CHANGELOG = await fs.readFile(mdFilePAth, 'utf8');

        const changelog = `${version} ${getDate(new Date())}\n\n${markdown}\n\n${CHANGELOG}`;

        await fs.writeFile(mdFilePAth, changelog, 'utf8');

        console.log(`${pkg} - changelog успешно записан`);
    } catch (error) {
        console.error('Ошибка при записи файла:', error);
    }
}
