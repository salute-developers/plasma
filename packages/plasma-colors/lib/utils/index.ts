import fs from 'fs';
import path from 'path';

// TODO: https://github.com/salute-developers/plasma/issues/1181

export const ROBO_COMMENT = '// Generated by robots, do not change this manually!\n';

export const generateFile = (name: string, content: string | object, deprecated = '') => {
    if (typeof content !== 'string') {
        content = JSON.stringify(content, null, 4);
    }

    return {
        file: `${name}.ts`,
        content: `export const ${name} = ${content};${deprecated}\n`,
    };
};

export type GeneratedFiles = Array<{
    file: string;
    content: string;
}>;

export const writeGeneratedToFS = (dir: string, generated: GeneratedFiles, generateMessage = ROBO_COMMENT) => {
    fs.existsSync(dir) || fs.mkdirSync(dir);

    for (const { file, content } of generated) {
        const src = path.join(dir, file);

        if (fs.existsSync(src)) {
            fs.appendFileSync(src, content);
            return;
        }

        fs.writeFileSync(src, `${generateMessage}\n${content}`);
    }
};

export const getDirectoryContent = (dir: string) =>
    fs
        .readdirSync(dir, { withFileTypes: true })
        .filter((item) => item.isFile() && path.extname(path.join(dir, item.name)) === '.json')
        .reduce((colorSchemas, item) => {
            const fileContent = fs.readFileSync(path.join(dir, item.name), 'utf-8');
            const name = path.parse(item.name).name;

            return {
                ...colorSchemas,
                [name]: fileContent,
            };
        }, {});
