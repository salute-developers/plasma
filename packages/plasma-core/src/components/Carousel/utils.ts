import { animatedScrollToX, animatedScrollToY, TimingFunction } from '../../utils';

import { ScrollAxis, ScrollAlign, UseCarouselLiteOptions } from './types';

const positionModByScrollAlign = ({
    scrollAlign,
    position,
    carouselSize,
    itemSize,
    offset,
    scrollStart,
    axis,
}: {
    scrollAlign: ScrollAlign;
    position: number;
    carouselSize: number;
    itemSize: number;
    offset: number;
    scrollStart: number;
    axis: ScrollAxis;
}) => {
    if (scrollAlign === 'start') {
        const inaccuracy = 1;
        const paddingOffset = axis === 'y' ? offset - itemSize / 2 + inaccuracy : 0;
        return position + paddingOffset;
    }
    if (scrollAlign === 'center') {
        return position - carouselSize / 2 + itemSize / 2;
    }
    if (scrollAlign === 'end') {
        return position - carouselSize + itemSize + offset;
    }
    if (scrollAlign === 'activeDirection') {
        if (position >= scrollStart + carouselSize - itemSize) {
            return position - carouselSize + itemSize + offset;
        }
        if (position > scrollStart) {
            return scrollStart;
        }
    }
    return position;
};

/**
 * Подсчет скролла до переданного индекса.
 */
export const getCalculatedPos = ({
    scrollEl,
    items,
    axis,
    index,
    offset,
    scrollAlign,
}: {
    scrollEl: HTMLElement;
    items: HTMLCollectionOf<HTMLElement>;
    axis: ScrollAxis;
    index: number;
    offset: number;
    scrollAlign: ScrollAlign;
}) => {
    let position = scrollAlign === 'center' ? offset : 0;
    let carouselSize;
    let itemSize;
    let scrollStart;

    if (items.item(index) === null) {
        return position;
    }

    for (let i = 0; i < index; i++) {
        if (axis === 'x') {
            position += items.item(i)?.offsetWidth ?? 0;
        } else {
            position += items.item(i)?.offsetHeight ?? 0;
        }
    }

    if (axis === 'x') {
        carouselSize = scrollEl.offsetWidth;
        itemSize = items.item(index)?.offsetWidth ?? 0;
        scrollStart = scrollEl.scrollLeft;
    } else {
        carouselSize = scrollEl.offsetHeight;
        itemSize = items.item(index)?.offsetHeight ?? 0;
        scrollStart = scrollEl.scrollTop;
    }

    return positionModByScrollAlign({
        scrollAlign,
        position,
        carouselSize,
        itemSize,
        offset,
        scrollStart,
        axis,
    });
};

/**
 * Подсчет смещения из-за паддингов.
 */
export const getCalculatedOffset = (scrollEl: Element, trackEl: Element, axis: ScrollAxis) => {
    const paddingProp = axis === 'x' ? 'paddingLeft' : 'paddingTop';
    return parseInt(getComputedStyle(scrollEl)[paddingProp], 10) + parseInt(getComputedStyle(trackEl)[paddingProp], 10);
};

/**
 * Прокрутка к указанной позиции с анимацией или без.
 */
export const scrollToPos = ({
    scrollEl,
    pos,
    axis,
    animated,
    duration,
    timingFunction,
}: {
    scrollEl: HTMLElement;
    pos: number;
    axis: ScrollAxis;
    animated?: boolean;
    duration?: number;
    timingFunction?: TimingFunction;
}) => {
    if (axis === 'x' && Math.abs(pos - scrollEl.scrollLeft) > 1) {
        if (animated) {
            animatedScrollToX(scrollEl, pos, duration, timingFunction);
        } else {
            scrollEl.scrollTo({ left: pos });
        }
    }
    if (axis === 'y' && Math.abs(pos - scrollEl.scrollTop) > 1) {
        if (animated) {
            animatedScrollToY(scrollEl, pos, duration, timingFunction);
        } else {
            scrollEl.scrollTo({ top: pos });
        }
    }
};

const round = (n: number) => Math.round(n * 100) / 100;

/**
 * Получить позицию (слот) айтема в каруселе.
 * Каждый айтем имеет свой слот относительно вьюпорта карусели.
 */
export const getItemSlot = (
    itemIndex: number,
    itemEnd: number,
    itemSize: number,
    scrollStart: number,
    scrollSize: number,
    scrollAlign: ScrollAlign,
    prevIndex = 0,
    offset = 0,
) => {
    /**
     * Граница и центр скролла (видимой части).
     * Смещение + размер.
     */
    const scrollEnd = scrollStart + scrollSize;
    const scrollCenter = scrollStart + scrollSize / 2;
    const itemCenter = itemEnd - itemSize / 2;

    if (scrollAlign === 'center') {
        return round((itemCenter - scrollCenter) / itemSize);
    }
    if (scrollAlign === 'start') {
        return round((itemEnd - itemSize - scrollStart) / itemSize);
    }
    if (scrollAlign === 'end') {
        return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
    }
    if (scrollAlign === 'activeDirection') {
        const prevStart = offset + itemSize * prevIndex;
        const prevEnd = prevStart + itemSize;
        const prevVisible = prevEnd > scrollStart && prevStart < scrollEnd;

        if (!prevVisible) {
            if (prevIndex < itemIndex) {
                return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
            }
            return round((itemEnd - itemSize - scrollStart) / itemSize);
        }
    }
    return null;
};

export function getCarouselItems(track: HTMLElement): HTMLCollectionOf<HTMLElement> {
    return track.children as HTMLCollectionOf<HTMLElement>;
}

const axisToTranslateMap: Record<UseCarouselLiteOptions['axis'], (position: number) => string> = {
    x(position) {
        return `translateX(${-position}px)`;
    },
    y(position) {
        return `translateY(${-position}px)`;
    },
};
const axisToOffsetKeyMap: Record<UseCarouselLiteOptions['axis'], 'offsetLeft' | 'offsetTop'> = {
    x: 'offsetLeft',
    y: 'offsetTop',
};

const axisToSizeKeyMap: Record<UseCarouselLiteOptions['axis'], 'offsetWidth' | 'offsetHeight'> = {
    x: 'offsetWidth',
    y: 'offsetHeight',
};

const axisToScrollKeyMap: Record<UseCarouselLiteOptions['axis'], 'scrollLeft' | 'scrollTop'> = {
    x: 'scrollLeft',
    y: 'scrollTop',
};

function getCenterPosition(itemSize: number, itemStart: number, carouselSize: number, trackOffset: number) {
    const relativeMiddle = itemStart + trackOffset + itemSize / 2;

    return relativeMiddle - carouselSize / 2;
}

function getEndPosition(itemSize: number, itemStart: number, carouselSize: number, trackOffset: number) {
    return 2 * trackOffset + itemStart + itemSize - carouselSize;
}

function getTranslatePosition(
    itemSize: number,
    itemStart: number,
    carouselSize: number,
    trackOffset: number,
    scrollAlign: UseCarouselLiteOptions['scrollAlign'],
): number {
    switch (scrollAlign) {
        case 'start': {
            return itemStart;
        }
        case 'end': {
            return getEndPosition(itemSize, itemStart, carouselSize, trackOffset);
        }
        case 'center':
        case 'activeDirection': // TODO: activeDirection сделать позже, мало кто использует. Fallback на 'center'
        default:
            return getCenterPosition(itemSize, itemStart, carouselSize, trackOffset);
    }
}

function boundPosition(position: number, trackSize: number, carouselSize: number, trackOffset: number): number {
    if (position < 0) {
        return 0;
    }

    // если все элементы помещаются в ширину карусели, то считаем trackEnd по-другому
    if (trackSize < carouselSize) {
        const trackEnd = Math.abs(carouselSize - trackSize - trackOffset);

        return position < trackEnd ? position : trackEnd;
    }

    const trackEnd = trackSize - carouselSize + trackOffset;

    if (position > trackEnd) {
        return trackEnd;
    }

    return position;
}

/**
 * Функция решает являются ли переданные индексы крайними с начала и конца.
 * Например в такой карусели `([1][2][3][4])` — при изменении индекса с 4 на 1 или наоборот функция вернёт `true`,
 * так как это переход между крайними элементами.
 *
 * Есть исключение, для карусели длинной в 2 элемента `([1][2])`.
 * В этом случае переход с индекса 1 на 2 и наоборот не должен считаться переходом с одного края на другой.
 *
 * @param prevIndex
 * @param index
 * @param numberOfItems количество элементов в карусели
 */
function isMaximumDistance(prevIndex: number, index: number, numberOfItems: number) {
    // проверка, что карусель состоит из 2-х элементов
    if (numberOfItems === 2) {
        return false;
    }

    return Math.abs(index - prevIndex) === numberOfItems - 1;
}

function setTimingFunction(element: HTMLElement | null, timingFunction: 'step-start' | '') {
    if (element) {
        element.style.transitionTimingFunction = timingFunction;
    }
}

function translateToPosition(element: HTMLElement, axis: UseCarouselLiteOptions['axis'], position: number): void {
    const translate = axisToTranslateMap[axis];

    element.style.transform = translate(position);
}

const scrollOptions: ScrollToOptions = { behavior: 'auto', left: 0, top: 0 };

/**
 * Делает расчет следующей позиции карусели исходя из параметров
 * index, align и размеров элементов track и carousel.
 * После применяет transform к элементу track или вызывает scrollTo на элементе carousel.
 * При scrollMode равном translate Анимирование происходит из-за CSS свойства `transition-property: transform`,
 * применённому к элементу track
 *
 * @param index индекс элемента к которому нужно сделать transform
 * @param prevIndex индекс предыдущего активного элемента для расчёта дистанции между индексами
 * @param axis ось вдоль которой будет происходить transform
 * @param align определяет позицию активного элемента относительно элемента carousel
 * @param track элемент к которому применяется transform
 * @param carousel элемент содержащий track
 * @param disableAnimation флаг для отключения анимирования прокрутки
 * @param scrollMode прокрутка через scrollTo или через translate
 */
export function translateToIndex(
    index: number,
    prevIndex: number,
    axis: UseCarouselLiteOptions['axis'],
    align: NonNullable<UseCarouselLiteOptions['scrollAlign']>,
    track: HTMLElement | null,
    carousel: HTMLElement | null,
    disableAnimation: boolean,
    scrollMode: NonNullable<UseCarouselLiteOptions['scrollMode']>,
): void {
    if (track === null || carousel === null) {
        return;
    }

    const scrollKey = axisToScrollKeyMap[axis];

    if (disableAnimation === false && scrollMode === 'translate' && carousel[scrollKey] !== 0) {
        carousel.scrollTo(scrollOptions);
    }

    const itemToTranslateTo = track.children.item(index) as HTMLElement | null;

    if (itemToTranslateTo === null) {
        return;
    }

    const numberOfItems = track.children.length;

    const offsetKey = axisToOffsetKeyMap[axis];
    const sizeKey = axisToSizeKeyMap[axis];

    const carouselSize = carousel[sizeKey];
    const trackSize = track[sizeKey];
    const trackOffset = track[offsetKey];
    const itemSize = itemToTranslateTo[sizeKey];
    const itemStart = itemToTranslateTo[offsetKey];

    const translatePosition = getTranslatePosition(itemSize, itemStart, carouselSize, trackOffset, align);

    const position = boundPosition(translatePosition, trackSize, carouselSize, trackOffset);

    if (scrollMode === 'scroll') {
        carousel.scrollTo({
            behavior: disableAnimation ? 'auto' : 'smooth',
            left: axis === 'x' ? position : 0,
            top: axis === 'y' ? position : 0,
        });

        return;
    }

    if (disableAnimation === true || isMaximumDistance(prevIndex, index, numberOfItems) === true) {
        // Выключаем стандартный easing, переключая его на step-start
        setTimingFunction(track, 'step-start');

        translateToPosition(track, axis, position);

        /**
         * Включаем стандартный easing.
         * нужно делать через setTimeout, чтобы transform успел произойти до изменения transitionTimingFunction
         */
        setTimeout(setTimingFunction, 0, track, '');

        return;
    }

    translateToPosition(track, axis, position);
}
