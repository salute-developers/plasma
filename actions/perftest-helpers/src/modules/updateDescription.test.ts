/* eslint-disable no-multi-assign, @typescript-eslint/no-explicit-any */
import { jest } from '@jest/globals';

import * as fsUtils from '../utils/fs';

import PerftestUpdateMergeRequestDescriptionApi from './updateDescription';

describe('perftest/updateDescription', () => {
    afterEach(() => {
        jest.clearAllMocks();
        jest.restoreAllMocks();
        jest.resetModules();
    });

    describe('update', () => {
        const prId = 12345;
        const reportPath = '/report/Path';
        const owner = 'salute-developers';
        const repo = 'plasma';

        let getUpdatedDescriptionMock: jest.Mock;
        let getMock: jest.Mock;
        let updateMock: jest.Mock;
        let readJsonMock: jest.SpiedClass<any>;

        class ApiWithStubbedInternals extends PerftestUpdateMergeRequestDescriptionApi {
            getUpdatedDescription = (getUpdatedDescriptionMock = jest.fn() as any);
        }

        let api: ApiWithStubbedInternals;

        beforeEach(() => {
            api = new ApiWithStubbedInternals({
                octokit: {
                    rest: {
                        pulls: {
                            get: getMock = jest.fn(),
                            update: updateMock = jest.fn(),
                        },
                    },
                } as any,
            });
            readJsonMock = jest.spyOn(fsUtils, 'readJson') as any;
        });

        it('should get current description and set new description', async () => {
            const prData = { data: { body: 'hehe' } } as any;
            const report = { result: {} };
            const basePrArgs = {
                owner,
                repo,
                pull_number: prId,
            };
            const updatedDescription = 'updated description';

            getMock.mockResolvedValue(prData as never);
            readJsonMock.mockResolvedValue(report);
            getUpdatedDescriptionMock.mockReturnValue(updatedDescription);
            updateMock.mockResolvedValue(null as never);

            await api.update({ prId, owner, repo, reportPath } as any);

            expect(getMock).toHaveBeenCalledTimes(1);
            expect(getMock).toHaveBeenCalledWith(basePrArgs);

            expect(readJsonMock).toHaveBeenCalledTimes(1);
            expect(readJsonMock).toHaveBeenCalledWith(reportPath);

            expect(getUpdatedDescriptionMock).toHaveBeenCalledTimes(1);
            expect(getUpdatedDescriptionMock).toHaveBeenCalledWith(prData.data.body, report);

            expect(updateMock).toHaveBeenCalledTimes(1);
            expect(updateMock).toHaveBeenCalledWith({ ...basePrArgs, body: updatedDescription });
        });
    });

    describe('getUpdatedDescription', () => {
        const surrouningsValue = { before: 'before', after: 'after' };
        const newReportValue = 'new report';
        let getSurroundingsMock: jest.Mock;
        let getVisualReportMock: jest.Mock;

        class ApiWithGetUpdatedDescriptionExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getUpdatedDescription = super.getUpdatedDescription;

            getSurroundings = (getSurroundingsMock = jest.fn().mockReturnValue(surrouningsValue) as any);

            getVisualReport = (getVisualReportMock = jest.fn().mockReturnValue(newReportValue) as any);
        }

        it('should get surroundings, get new report and combine them', () => {
            const currentDescription = 'current';
            const jsonReport = { result: {} };
            const res = new ApiWithGetUpdatedDescriptionExposed({} as any).getUpdatedDescription(
                currentDescription,
                jsonReport as any,
            );

            expect(getSurroundingsMock).toHaveBeenCalledTimes(1);
            expect(getSurroundingsMock).toHaveBeenCalledWith(currentDescription);
            expect(getVisualReportMock).toHaveBeenCalledTimes(1);
            expect(getVisualReportMock).toHaveBeenCalledWith(jsonReport);

            expect(res).toEqual([surrouningsValue.before, newReportValue, surrouningsValue.after].join(''));
        });
    });

    describe('getSurroundings', () => {
        const wrapContentImplementation = (r: string) => `<div>${r}</div>`;
        let wrapContentMock: jest.Mock;
        class ApiWithGetSurroundingsExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getSurroundings = super.getSurroundings;

            wrapReportContent = (wrapContentMock = jest
                .fn()
                .mockImplementation(wrapContentImplementation as any)) as any;
        }

        it('should return exact before and after text if report exists in text', () => {
            const api = new ApiWithGetSurroundingsExposed({} as any);
            const realBefore = 'blabla\n\nblabla';
            const realAfter = 'blabla\nwerwqet';
            const realReport = wrapContentImplementation('immareport');
            const mrDescription = `${realBefore}${realReport}${realAfter}`;

            const { before, after } = api.getSurroundings(mrDescription);

            expect(before).toEqual(realBefore);
            expect(after).toEqual(realAfter);
            expect([before, realReport, after].join('')).toEqual(mrDescription);

            expect(wrapContentMock).toHaveBeenCalledTimes(1);
        });

        it('should return before canary release text and everything after if no report found', () => {
            const api = new ApiWithGetSurroundingsExposed({} as any);
            const realBefore = 'blabla\n\nblabla';
            const realAfter = '<!-- GITHUB_RELEASE -->\nblabla\nwerwqet\nblabla\nkaoitwlek';
            const mrDescription = `${realBefore}${realAfter}`;

            const { before, after } = api.getSurroundings(mrDescription);

            expect(before).toEqual(`${realBefore}\n\n`);
            expect(after).toEqual(`\n\n${realAfter}`);

            expect(wrapContentMock).toHaveBeenCalledTimes(1);
        });

        it('should return all the text as before if no report found and no section found', () => {
            const api = new ApiWithGetSurroundingsExposed({} as any);
            const mrDescription = 'blabla\n\nblabla\nblabla\nwerwqet\nblabla\nkaoitwlek';

            const { before, after } = api.getSurroundings(mrDescription);

            expect(before).toEqual(`${mrDescription}\n\n`);
            expect(after).toEqual('\n\n');

            expect(wrapContentMock).toHaveBeenCalledTimes(1);
        });
    });

    describe('getVisualReport', () => {
        const descriptionResult = 'description';
        let getDescriptionMock: jest.Mock;
        let wrapContentMock: jest.Mock;

        class ApiWithGetVisualReportExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getVisualReport = super.getVisualReport;

            getReportDescription = (getDescriptionMock = jest.fn().mockReturnValue(descriptionResult) as any);

            wrapReportContent = (wrapContentMock = jest.fn().mockImplementation((r) => r) as any);
        }

        it('should create report markup, add description and wrap', () => {
            const jsonReport = { hasSignificantNegativeChanges: true };
            const api = new ApiWithGetVisualReportExposed({} as any);

            const result = api.getVisualReport(jsonReport as any);

            expect(result.includes('FAIL')).toBe(true);

            expect(getDescriptionMock).toHaveBeenCalledTimes(1);
            expect(getDescriptionMock).toHaveBeenCalledWith(jsonReport);

            expect(wrapContentMock).toHaveBeenCalledTimes(1);
            expect((wrapContentMock.mock.calls[0][0] as any).includes(descriptionResult)).toBe(true);
            expect((wrapContentMock.mock.calls[0][0] as any).includes('FAIL')).toBe(true);
        });

        it('should create OK report if no significant negative changes', () => {
            const jsonReport = { hasSignificantNegativeChanges: false };
            const api = new ApiWithGetVisualReportExposed({} as any);

            const result = api.getVisualReport(jsonReport as any);
            expect(result.includes('OK')).toBe(true);
        });
    });

    describe('getReportDescription', () => {
        const jsonReport = { result: {} };
        const resultByTaskId = { kek: 'bla' };

        let getResultByTaskIdMock: jest.Mock;
        let getDescriptionTablesMock: jest.Mock;

        class ApiWithGetReportDescriptionExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getReportDescription = super.getReportDescription;

            getResultByTaskId = (getResultByTaskIdMock = jest.fn().mockReturnValue(resultByTaskId) as any);

            getDescriptionTables = (getDescriptionTablesMock = jest.fn() as any);
        }

        it('should return empty description if no significant changes to show', () => {
            const api = new ApiWithGetReportDescriptionExposed({} as any);

            getDescriptionTablesMock.mockReturnValue('');

            const res = api.getReportDescription(jsonReport as any);

            expect(res).toEqual('');

            expect(getResultByTaskIdMock).toHaveBeenCalledTimes(1);
            expect(getResultByTaskIdMock).toHaveBeenCalledWith(jsonReport);

            expect(getDescriptionTablesMock).toHaveBeenCalledTimes(1);
            expect(getDescriptionTablesMock).toHaveBeenCalledWith(resultByTaskId);
        });

        it('should return description markup with report', () => {
            const api = new ApiWithGetReportDescriptionExposed({} as any);
            const report = 'report';

            getDescriptionTablesMock.mockReturnValue(report);

            const res = api.getReportDescription(jsonReport as any);

            expect(res).toMatchSnapshot();

            expect(getResultByTaskIdMock).toHaveBeenCalledTimes(1);
            expect(getResultByTaskIdMock).toHaveBeenCalledWith(jsonReport);

            expect(getDescriptionTablesMock).toHaveBeenCalledTimes(1);
            expect(getDescriptionTablesMock).toHaveBeenCalledWith(resultByTaskId);
        });
    });

    describe('getResultByTaskId', () => {
        class ApiWithGetResultByTaskIdExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getResultByTaskId = super.getResultByTaskId;
        }

        const api = new ApiWithGetResultByTaskIdExposed({} as any);

        it('should return taskId to [subjectId, compareResult][] map with significantly changed metrics', () => {
            const jsonReport = {
                result: {
                    subjectId1: {
                        taskId1: { mean: { change: { percentage: 1, significanceRank: 'high' } } },
                        taskId2: { mean: { change: { percentage: 4, significanceRank: 'high' } } },
                        taskId3: { mean: { change: { percentage: 5, significanceRank: 'high' } } },
                        taskIdFiltered1: { mean: { change: { percentage: 0.5, significanceRank: 'low' } } },
                        taskIdFiltered2: { mean: { new: {} } },
                    },
                    subjectId2: {
                        taskId1: { mean: { change: { percentage: 2, significanceRank: 'high' } } },
                        taskId2: { mean: { change: { percentage: 3, significanceRank: 'high' } } },
                    },
                    subjectId3: {
                        taskIdFiltered1: { mean: { change: { percentage: 0.5, significanceRank: 'low' } } },
                        taskIdFiltered2: { mean: { new: {} } },
                    },
                },
            };

            const result = api.getResultByTaskId(jsonReport as any);

            expect(new Set(Object.keys(result))).toEqual(new Set(['taskId1', 'taskId2', 'taskId3']));

            expect(result.taskId1).toEqual(
                (['subjectId2', 'subjectId1'] as const).map((id) => [id, jsonReport.result[id].taskId1.mean]),
            );

            expect(result.taskId2).toEqual(
                (['subjectId1', 'subjectId2'] as const).map((id) => [id, jsonReport.result[id].taskId2.mean]),
            );

            expect(result.taskId3).toEqual(
                (['subjectId1'] as const).map((id) => [id, jsonReport.result[id].taskId3.mean]),
            );
        });
    });

    describe('getDescriptionTables', () => {
        let getSubjectTaskResultTableRowMock: jest.Mock;

        class ApiWithGetMetricsReportExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getDescriptionTables = super.getDescriptionTables;

            public getSubjectTaskResultTableRow = (getSubjectTaskResultTableRowMock = jest.fn() as any);
        }

        it('should iterate through tasks, forming tables with rows', () => {
            const api = new ApiWithGetMetricsReportExposed({} as any);
            const resultByTaskId = {
                taskId1: [
                    ['subjectId1', {}],
                    ['subjectId2', {}],
                ],
                taskId2: [
                    ['subjectId1', {}],
                    ['subjectId2', {}],
                    ['subjectId3', {}],
                ],
                taskId3: [['subjectId1', {}]],
            };
            getSubjectTaskResultTableRowMock.mockReturnValue('row');

            const res = api.getDescriptionTables(resultByTaskId as any);

            expect(res).toMatchSnapshot();
        });
    });

    describe('getSubjectTaskResultTableRow', () => {
        let formatValueMock: jest.Mock;

        class ApiWithGetSubjectTaskResultTableRowExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public getSubjectTaskResultTableRow = super.getSubjectTaskResultTableRow;

            public formatValue = (formatValueMock = jest.fn().mockImplementation((r) => r) as any);
        }

        it('should format the row', () => {
            const api = new ApiWithGetSubjectTaskResultTableRowExposed({} as any);
            const res = api.getSubjectTaskResultTableRow('subjectId', {
                old: 'old',
                new: ['new1', 'new2'],
                change: {
                    difference: 'diff',
                    percentage: 'percentage',
                },
            } as any);

            expect(res).toMatchSnapshot();

            expect(formatValueMock).toHaveBeenCalledTimes(4);
        });
    });

    describe('formatValue', () => {
        class ApiWithWrapFormatValueExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public formatValue = super.formatValue;
        }

        const { formatValue } = new ApiWithWrapFormatValueExposed({} as any);

        it('should format number with fixed precision', () => {
            expect(formatValue(Math.PI)).toEqual('3.14');
            expect(formatValue(Math.PI, { fractionDigits: 4 })).toEqual('3.1416');
        });

        it('should format number with suffix', () => {
            expect(formatValue(Math.PI, { suffix: 'pts' })).toEqual('3.14pts');
        });

        it('should format number with sign', () => {
            expect(formatValue(-Math.PI, { sign: false })).toEqual('-3.14');
            expect(formatValue(Math.PI, { sign: true })).toEqual('+3.14');
            expect(formatValue(-Math.PI, { sign: true })).toEqual('-3.14');
            expect(formatValue(0, { sign: true, fractionDigits: 0 })).toEqual('0');
        });

        it('should format number with highlighter', () => {
            expect(formatValue(Math.PI, { highlight: true })).toEqual('{+(3.14)+}');
            expect(formatValue(-Math.PI, { highlight: true })).toEqual('{-(-3.14)-}');
            expect(formatValue(0, { highlight: true })).toEqual('`(0.00)`');
        });
    });

    describe('wrapReportContent', () => {
        class ApiWithWrapContentExposed extends PerftestUpdateMergeRequestDescriptionApi {
            public wrapReportContent = super.wrapReportContent;
        }

        it('should wrap content string', () => {
            const content = 'asd';
            const res = new ApiWithWrapContentExposed({} as any).wrapReportContent(content);

            expect(res.includes(content)).toBe(true);
            expect(res).toMatchSnapshot();
        });
    });
});
