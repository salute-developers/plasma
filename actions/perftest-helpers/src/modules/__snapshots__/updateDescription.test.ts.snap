// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`perftest/updateDescription getDescriptionTables should iterate through tasks, forming tables with rows 1`] = `
"
**taskId1:**

| Компонент | Изменение | Было | Стало |
|:-:|:-:|:-:|:-:|
row
row

**taskId2:**

| Компонент | Изменение | Было | Стало |
|:-:|:-:|:-:|:-:|
row
row
row

**taskId3:**

| Компонент | Изменение | Было | Стало |
|:-:|:-:|:-:|:-:|
row
"
`;

exports[`perftest/updateDescription getReportDescription should return description markup with report 1`] = `
"<details><summary>Details</summary><br/>
report
</details>"
`;

exports[`perftest/updateDescription getSubjectTaskResultTableRow should format the row 1`] = `"| subjectId | diff percentage | old | new1 |"`;

exports[`perftest/updateDescription wrapReportContent should wrap content string 1`] = `
"<!-- AUTO-GENERATED PERFTEST REPORT -->
asd
<!-- AUTO-GENERATED PERFTEST REPORT -->"
`;
